bin/python/ripple/build/module/Default.py

```diff
@@ -0,0 +1,215 @@
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals)
+
+import os
+
+from beast.build.Module import Module
+
+from beast.build.Function import Env, compose, for_tags, not_tags
+
+Append = Env.Append
+Prepend = Env.Prepend
+
+
+MODULE = Module(
+    files=compose(
+        for_tags(
+            'debug',
+            Append(CPPDEFINES=['DEBUG', '_DEBUG']),
+        ),
+
```

Comment:
*Complexity*

Now, these are excellent points, each of which is an additional complexity
- and as you know I consider complexity a huge threat. (I can also fix some
of these issues...)

That said, there is a great deal of complexity in what we are requiring of
our build system, all of which has to go somewhere.


*Four phase build.*

Let's start with the "four phase build"
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/Module.py#L14-L26>
-
before, files, target, after.

We have main targets like rippled, and within them build variants like
gcc.debug and this structure is an existing precondition that can't be
changed.

This requires us to have two different build states:  we have one BuildState
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/BuildState.py>,
but multiple BuildVariants
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/BuildVariant.py>
.

Given the constraints, there isn't another way to do it.


*Functional programming*

Now, why are we returning functions everywhere?  Why not just... do things?

Two reasons:

The first is that Scons tends to work that way - you prepare a description
of what's going to happen, and then Scons does it.

The second is that I wanted Modules, as much as possible, to be expressible
as static data - something that can be constructed at program loading time
and never changed - because this makes reuse and reference so much easier.
But at the time the Python program loads, we simply don't have any build
information ready!

(Having separate Modules we can reuse was one of the stated goals of the
project...)



*Hard to navigate directories.*

As for the complexity of having code split over two separate directories
without a common root - tell me about it.  :-(  This is an unavoidable
consequence of having the segmentation between "beast" and "ripple".

I tried to make the hierarchies exactly the same:


    bin/python/ripple
    ├── build
    │   ├── Config.py
    │   ├── module
    │   │   ├── Default.py
    │   │   └── __init__.py
    │   └── Unity.py

    src/beast/python/beast
    ├── build
    │   ├── BuildState.py
    │   ├── BuildVariant.py
    │   ├── CcWarning.py
    │   ├── CheckForFastBuild.py
    │   ├── Config.py
    │   ├── Function.py
    │   ├── Git.py
    │   ├── module
    │   │   ├── Beast.py
    │   │   ├── Boost.py
    │   │   ├── CountTests.py
    │   │   ├── GitTag.py
    │   │   ├── Jemalloc.py
    │   │   ├── MSVC.py
    │   │   ├── OpenSSL.py
    │   │   ├── PhonyTargets.py
    │   │   ├── Proto.py
    │   │   └── Vcxproj.py
    │   ├── Module.py
    │   ├── TagSet.py
    │   ├── Toolchain.py
    │   └── Util.py
    ├── Color.py
    └── System.py


(This graph omits test code, and empty __init__.py files.)



*Maintenance*

Big issue for me!

What is likely to happen in future maintenance?   Here's my guess, by
decreasing frequency:

   1. We're going to add a lot of Modules.
   2. We'll add new binary targets: e.g. tests.
   3. We'll add new options like --help (or replacing --ninja).
   4. We'll occasionally have to add a new "build phase", or fix details in
   an existing build phase.
   5. We'll make existential overhauls to the whole "tags and variants"
   idea.


So a top priority was for me *making Modules easy to write and maintain.*

And - I think - they are.

It's really easy to create a module that just has source code
<https://github.com/rec/rippled/blob/scons/bin/python/ripple/build/module/__init__.py#L33-L38>
.

There's a special case
<https://github.com/rec/rippled/blob/scons/bin/python/ripple/build/module/__init__.py#L89-L106>
for unity builds if the unity file has the same name as the directory - but
we can also handle complex cases like the "legacy build"
<https://github.com/rec/rippled/blob/scons/bin/python/ripple/build/module/__init__.py#L43-L64>
.


Now, some of our modules have fairly complex logic.  For example, for
boost, there's a list of hard-compiled libraries, and we don't require
BOOST_ROOT to be set, but if it is, we need to explicitly load from there,
and to prefer the static libraries if possible - here
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/module/Boost.py>
.



On Tue, Aug 25, 2015 at 4:48 PM, seelabs <notifications@github.com> wrote:

> In bin/python/ripple/build/module/Default.py
> <https://github.com/ripple/rippled/pull/1252#discussion_r37917453>:
>
> > +
> > +from beast.build.Module import Module
> > +
> > +from beast.build.Function import Env, compose, for_tags, not_tags
> > +
> > +Append = Env.Append
> > +Prepend = Env.Prepend
> > +
> > +
> > +MODULE = Module(
> > +    files=compose(
> > +        for_tags(
> > +            'debug',
> > +            Append(CPPDEFINES=['DEBUG', '_DEBUG']),
> > +        ),
> > +
>
> @rec <https://github.com/rec>: I really like the interface you have here.
> Very nice. However, I don't like the complexity needed to support this
> interface. I can't do better, but I'm also concerned about anyone on the
> team other than you supporting this python code. Looking at the
> implementation, there are lots of functions passed around. It makes the
> code hard to follow for someone (like me) who isn't used to the style. For
> example:
>
> 1) Look at what for_tags does. This is defined in Function.py
>
> 2) That calls: _tagger with the items and a matcher.
>
> 3) _tagger returns a function that iterates through all the callable
> objects that pass the matcher with the non-callable objects (I did a
> double-take on (functions.append if callable(i) else tags.add)(i))
>
> 4) Append is Env.Append. Env is defined in Function.py as EnvClass.
> EnvClass doesn't have an attribute Append, but it does define a
> __getattr__ function that return a function that returns a function (two
> levels) that calls a function on the build env.
>
> 5) compose is defined in Function.py and returns a function that calls
> all of the arguments originally passed to compose.
>
> 6) The Module __init__ function looks at the files parameter. If it's
> callable, it saves it, if not there are other special cases.
>
> I'm also having some problems navigating all the python files that support
> our build system.
>
> I don't know how actionable that feedback is, but I did want to share the
> concern.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ripple/rippled/pull/1252/files#r37917453>.
>